cpp/include/raft/spatial/knn/detail/ann_utils.cuh

```diff
@@ -0,0 +1,491 @@
+ * row was selected along the way.
+ *
+ * @tparam T
+ *
+ * @param n_cols number of columns in all matrices
+ * @param[out] output output matrix [..., n_cols]
+ * @param[out] selection_counters number of occurrences of each row id in row_ids [..., n_cols]
+ * @param n_rows number of rows in the input
+ * @param[in] input row-major input matrix [n_rows, n_cols]
+ * @param[in] row_ids row indices in the output matrix [n_rows]
+ */
+template <typename T>
+void accumulate_into_selected(uint32_t n_rows,
+                              uint32_t n_cols,
+                              float* output,
+                              uint32_t* selection_counters,
+                              const T* input,
+                              const uint32_t* row_ids,
+                              rmm::cuda_stream_view stream)
+{
+  switch (check_pointer_residency(output, input, selection_counters, row_ids)) {
+    case pointer_residency::host_and_device:
+    case pointer_residency::device_only: {
+      uint32_t block_dim = 128;
+      auto grid_dim      = static_cast<uint32_t>(ceildiv<uint64_t>(
+        static_cast<uint64_t>(n_rows) * static_cast<uint64_t>(n_cols), block_dim));
+      accumulate_into_selected_kernel<T><<<grid_dim, block_dim, 0, stream>>>(
+        n_rows, n_cols, output, selection_counters, input, row_ids);
+    } break;
+    case pointer_residency::host_only: {
+      stream.synchronize();
+      for (uint64_t i = 0; i < n_rows; i++) {
+        uint64_t l = row_ids[i];
+        selection_counters[l]++;
+        for (uint64_t j = 0; j < n_cols; j++) {
+          output[j + n_cols * l] += mapping<float>{}(input[j + n_cols * i]);
+        }
+      }
+      stream.synchronize();
+    } break;
+    default: RAFT_FAIL("All pointers must reside on the same side, host or device.");
+  }
+}
+
+__global__ void normalize_rows_kernel(uint32_t n_rows, uint32_t n_cols, float* a)
+{
+  uint64_t i = threadIdx.y + (blockDim.y * blockIdx.x);
+  if (i >= n_rows) return;
+
+  float sqsum = 0.0;
+  for (uint32_t j = threadIdx.x; j < n_cols; j += blockDim.x) {
+    float val = a[j + (n_cols * i)];
+    sqsum += val * val;
+  }
+  sqsum += __shfl_xor_sync(0xffffffff, sqsum, 1);
+  sqsum += __shfl_xor_sync(0xffffffff, sqsum, 2);
+  sqsum += __shfl_xor_sync(0xffffffff, sqsum, 4);
+  sqsum += __shfl_xor_sync(0xffffffff, sqsum, 8);
+  sqsum += __shfl_xor_sync(0xffffffff, sqsum, 16);
+  if (sqsum <= 1e-8) return;
+  sqsum = rsqrtf(sqsum);  // reciprocal of the square root
+  for (uint32_t j = threadIdx.x; j < n_cols; j += blockDim.x) {
+    a[j + n_cols * i] *= sqsum;
+  }
+}
+
+/**
+ * @brief Divide rows by their L2 norm (square root of sum of squares).
+ *
+ * NB: device-only function
+ *
+ * @param[in] n_rows
+ * @param[in] n_cols
+ * @param[inout] a device pointer to a row-major matrix [n_rows, n_cols]
+ * @param stream
+ */
+inline void normalize_rows(uint32_t n_rows, uint32_t n_cols, float* a, rmm::cuda_stream_view stream)
```

Comment:
@achirkin I think it's ok, I don't really see a case where one would compute the norm / normalize a thick matrix, since the rows are typically samples and the cols their dimensionality. In fact, if you look at `raft::linalg::rowNorm`, it is using the `coalescedReductionKernel` which is doing a cub-based reduction with one block per row, quite similarly to `normalize_rows_kernels`.

I'll replace the hardcoded reduction in `normalize_rows_kernel` with a cub-based reduction and move it to the matrix or linalg namespace (and benchmark before/ater perf)